katello/hooks/pre/32-cdn_setting.rb

```diff
@@ -0,0 +1,10 @@
+if module_enabled?('katello')
+  if param_value('katello', 'cdn_ssl_version')
```

Comment:
From all of my testing, this won't be available here if it's been dropped from the puppet module. You'll have to instead take a tactic of using a `pre_commit` hook and accessing `@kafo.config.answers` which will have the answers prior to being committed to disk and being overwritten.

Reply:
@ehelms if we drop the puppet code will the `/etc/foreman/settings.d/katello.yaml` get updated with the removal of the cdn_ssl_setting or will we have to manually remove it from that file directly?

Reply:
AFAIK Kafo doesn't drop unknown answers and just keeps them. If the puppet-katello module doesn't have the parameter, you don't need the whole pre hook, call foreman-rake in a post hook and then delete the answer.

Reply:
That is what I was planning on going with @ekohl but then I saw the comment from Eric.

Reply:
My experimentation has shown that any parameters that are unknown to Kafo will not be propagated to the answers file when Kafo commits the parameters from memory to the file. Thus having to do this in `pre_commit` was how we solved it in another case.

Reply:
Interesting, so https://github.com/theforeman/foreman-installer/blob/develop/config/foreman.migrations/20190111180118_delete_removed_settings.rb was mostly redundant?

However, with pre_commit I wonder if that works. Can you actually create a setting using `foreman-rake config` or do you need to do it after running the migrations to ensure it's present (as I've been assuming)?

Reply:
foreman-rake config creates a setting in the Foreman database, I'm missing why that matters when its ran.

Reply:
https://github.com/theforeman/foreman/blob/26dda8dbbf32d357657c3d22018c9d247fb0cb4e/lib/tasks/config.rake#L102

It will hard fail if the database row with that settings name doesn't exist:
```console
# foreman-rake -- config -k doesnotexist -v foo
rake aborted!
NoMethodError: undefined method `settings_type' for nil:NilClass
/usr/share/foreman/lib/tasks/config.rake:105:in `block in run_key_values'
/usr/share/foreman/lib/tasks/config.rake:100:in `each'
/usr/share/foreman/lib/tasks/config.rake:100:in `run_key_values'
/usr/share/foreman/lib/tasks/config.rake:77:in `run'
/usr/share/foreman/lib/tasks/config.rake:167:in `block in <top (required)>'
/opt/rh/rh-ruby25/root/usr/share/gems/gems/rake-12.3.0/exe/rake:27:in `<top (required)>'
Tasks: TOP => config
(See full trace by running task with --trace)
```

Usually the database migrations take care of creating this, hence why it should be a post hook. Otherwise you're creating an upgrade issue.

Reply:
The data needs to be stored somewhere "temporarily" I think. Whether thats a file that gets cleaned up at the end of post section or on the scenario configuration. As we want to consider if the installer fails we don't want to have erased the data. So this could be the hook creating it or some more supported entity where old parameters are stored.

Reply:
That was my thinking as well. I had hoped we could use the answers file but looks like we should use the [store](https://github.com/theforeman/kafo#store) (`store_custom_config` and `get_custom_config`). I don't exactly know at which stage the `store_custom_config` method should be called so you'll need to experiment with that.